### PR TITLE
New function to remove challenge in dojos.py

### PR DESCRIPTION
**Pull request by wumingzhilian**
New function to remove challenge in dojos.py

### CODE REVIEW

**Comment by zardus** [comment]
This functionality scares me because of the permanence... What do you think about gating it behind `is_admin()` so that only ctfd admins are about to do it?

**Comment by wumingzhilian** [comment]
You are right, as I don't know the dojo code very well yet, I just used the dojo.is_admin() function in the code, which I think will satisfy the admin user for deletion.

**Comment by zardus** [comment]
Yeah, that just makes sure that the user is an admin of the *dojo*, which eventually will probably be sufficient, but for now, let's make that a global `is_admin()`.

**Comment by ConnorNelson** [comment]
I am also scared about deleting dojos, this sounds like a potential email saying "I accidentally deleted my dojo, please help". I like to avoid potential emails.

Is there a use-case for this? I would much rather something that "deactivates", whether that simply hides it from the dojos page, or even 404s all related dojo pages for non-dojo-admins.

If we ever want to *actually* delete a dojo, I am fine declaring that this should be a manual database operation.

**Comment by zardus** [comment]
Gating it behind is_admin is roughly the same as a database operation, right? Just a bit less prone to errors from us...

**Comment by zardus** [comment]
BTW, @wumingzhilian, could you add a test case for this?

**Comment by rileyannon** [comment]
For most places that have permanent deletion like this, I often see one of two things:

1. 30 day restoration period
2. Confirmation via email/MFA
3. A way to download data so you can easily restore it if you need so

But I'm just a random bystander so just my two cents on this ¯\_(ツ)_/¯

**Comment by zardus** [comment]
Yeah, that sort of complexity is exactly what I'd like to avoid by making this admin-only :-)

At least in terms of the last one, since all dojos are external git repos anyways, we don't have to worry about it.

**Comment by mudongliang** [comment]
@wumingzhilian is my student. We developed this functionability due to the inconvenience during our downstream testing. Actually we only delete dojos that are in our testing deployment. This is a really dangerous action in the production environment.

Maybe we can limit this deletion functionability to only dojos that have no user activities. Then actual data deletion can be only done by database operation.

**Comment by wumingzhilian** [comment]
> Gating it behind is_admin is roughly the same as a database operation, right? Just a bit less prone to errors from us...

Yes, the function is the same.

**Comment by wumingzhilian** [comment]
> Yeah, that just makes sure that the user is an admin of the _dojo_, which eventually will probably be sufficient, but for now, let's make that a global `is_admin()`.

Yes, . I misunderstood the functionality of the dojo.is_admin() function. Deleting content from the dojo should indeed be restricted to site administrators. Is there a convenient function in the dojo for me to call for this purpose? I'm just a novice developer :)))

**Comment by wumingzhilian** [comment]
> For most places that have permanent deletion like this, I often see one of two things:
> 
> 1. 30 day restoration period
> 2. Confirmation via email/MFA
> 3. A way to download data so you can easily restore it if you need so
> 
> But I'm just a random bystander so just my two cents on this ¯\_(ツ)_/¯

Thank you for your suggestion.

**Comment by zardus** [comment]
`dojo.is_admin` checks if the current user is an admin of that specific dojo, CTFd's `is_admin` checks that the user is an admin site-wide in CTFd. We should use the latter here, add a testcase, and then I think this PR is good to go!

Long-term, we should probably rename the concept of dojo admin to something else to avoid the confusion. Maybe `dojo.is_sensei`? That's a different topic, though.

BTW, if you rebase this branch on master, I think I should be able to enable testcases to run on the PR.

**Comment by wumingzhilian** [comment]
> `dojo.is_admin` checks if the current user is an admin of that specific dojo, CTFd's `is_admin` checks that the user is an admin site-wide in CTFd. We should use the latter here, add a testcase, and then I think this PR is good to go!
> 
> Long-term, we should probably rename the concept of dojo admin to something else to avoid the confusion. Maybe `dojo.is_sensei`? That's a different topic, though.
> 
> BTW, if you rebase this branch on master, I think I should be able to enable testcases to run on the PR.

I have successfully tested the new pull request locally. Only the CTFd admin user can see the delete function, but I might not know how to write it for CI, so I may not be able to help much. sorry~~

File: dojo_plugin/pages/dojos.py [hunk 1]
```diff
@@ -112,6 +112,32 @@ def update_dojo(dojo, update_code=None):
         return {"success": False, "error": str(e)}, 400
     return {"success": True}
 
+@dojos.route("/dojo/<dojo>/delete/", methods=["GET","POST"])
+@dojos.route("/dojo/<dojo>/delete/<update_code>", methods=["GET","POST"])
```
[the remark this answers is not in the document]
Comment (zardus) [reply dojo_plugin/pages/dojos.py:116]: I see. It's modeled on `update_dojo`. I think for this functionality, let's make it POST-only, with CSRF protections, and not use the `update_code` here. We'll rely on the `is_admin` check, which is sufficient here (with CSRF protections).

File: dojo_theme/templates/dojo_admin.html [hunk 2]
```diff
@@ -28,6 +28,16 @@ <h2>{{ dojo.name }}</h2>
       </a>
       </figure>
     </div>
+    {% if is_admin() %}
+    <div class="col-lg-auto m-3">
+      <figure>
+        <a class="text-decoration-none" href="{{ url_for('pwncollege_dojos.delete_dojo', dojo=dojo.reference_id,update_code=dojo.update_code) }}" target="_blank">
```
Comment (zardus) [inline dojo_theme/templates/dojo_admin.html:34]: As mentioned in another comment, let's make this a POST form similar to `dojo-promote-dojo`.

**Review by zardus — CHANGES_REQUESTED** [review]
Testcase is super easy, something like this:

```
diff --git a/test/test_running.py b/test/test_running.py
index 20105b1..d470eef 100644
--- a/test/test_running.py
+++ b/test/test_running.py
@@ -9,7 +9,7 @@ import pytest
 
 #pylint:disable=redefined-outer-name,use-dict-literal,missing-timeout,unspecified-encoding,consider-using-with
 
-from utils import PROTO, HOST, login, dojo_run, workspace_run
+from utils import PROTO, HOST, login, dojo_run, workspace_run, create_dojo_yml
 
 def get_flag(user):
     return workspace_run("cat /flag", user=user, root=True).stdout
@@ -87,6 +87,14 @@ def test_create_import_dojo(example_import_dojo, admin_session):
 def test_start_challenge(admin_session):
     start_challenge("example", "hello", "apple", session=admin_session)
 
+@pytest.mark.dependency(depends=["test_create_dojo"])
+def test_delete_dojo(admin_session):
+    reference_id = create_dojo_yml("""id: delete-test""", session=admin_session)
+    print(reference_id)
+    assert admin_session.get(f"{PROTO}://{HOST}/{reference_id}/").status_code == 200
+    assert admin_session.get(f"{PROTO}://{HOST}/dojo/{reference_id}/delete/").status_code == 200
+    assert admin_session.get(f"{PROTO}://{HOST}/{reference_id}/").status_code == 404
+
 @pytest.mark.dependency(depends=["test_create_dojo"])
 def test_join_dojo(admin_session, guest_dojo_admin):
     random_user_name, random_session = guest_dojo_admin
```

Other than that, I'll mark a few final requests and we can get this thing going!

File: dojo_theme/templates/dojo_admin.html (hunk 2 above)
Comment (wumingzhilian) [reply dojo_theme/templates/dojo_admin.html]: okay，now delete function is POST

File: dojo_plugin/pages/dojos.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (wumingzhilian) [reply dojo_plugin/pages/dojos.py]: update_code is deleted

File: dojo_plugin/pages/dojos.py [hunk 3]
```diff
@@ -112,6 +112,32 @@ def update_dojo(dojo, update_code=None):
...
+@bypass_csrf_protection
```
[the remark this answers is not in the document]
Comment (wumingzhilian) [reply dojo_plugin/pages/dojos.py]: sorrrrrrry,I've changed it

**Comment by wumingzhilian** [comment]
> Testcase is super easy, something like this:
> 
> ```
> diff --git a/test/test_running.py b/test/test_running.py
> index 20105b1..d470eef 100644
> --- a/test/test_running.py
> +++ b/test/test_running.py
> @@ -9,7 +9,7 @@ import pytest
>  
>  #pylint:disable=redefined-outer-name,use-dict-literal,missing-timeout,unspecified-encoding,consider-using-with
>  
> -from utils import PROTO, HOST, login, dojo_run, workspace_run
> +from utils import PROTO, HOST, login, dojo_run, workspace_run, create_dojo_yml
>  
>  def get_flag(user):
>      return workspace_run("cat /flag", user=user, root=True).stdout
> @@ -87,6 +87,14 @@ def test_create_import_dojo(example_import_dojo, admin_session):
>  def test_start_challenge(admin_session):
>      start_challenge("example", "hello", "apple", session=admin_session)
>  
> +@pytest.mark.dependency(depends=["test_create_dojo"])
> +def test_delete_dojo(admin_session):
> +    reference_id = create_dojo_yml("""id: delete-test""", session=admin_session)
> +    print(reference_id)
> +    assert admin_session.get(f"{PROTO}://{HOST}/{reference_id}/").status_code == 200
> +    assert admin_session.get(f"{PROTO}://{HOST}/dojo/{reference_id}/delete/").status_code == 200
> +    assert admin_session.get(f"{PROTO}://{HOST}/{reference_id}/").status_code == 404
> +
>  @pytest.mark.dependency(depends=["test_create_dojo"])
>  def test_join_dojo(admin_session, guest_dojo_admin):
>      random_user_name, random_session = guest_dojo_admin
> ```
> 
> Other than that, I'll mark a few final requests and we can get this thing going!

now ,regarding ci's function is complete(

File: dojo_plugin/pages/dojos.py [hunk 4]
```diff
@@ -124,7 +145,7 @@ def view_dojo_admin(dojo):
 @dojos.route("/dojo/<dojo>/admin/activity")
 @dojo_route
 def view_dojo_activity(dojo):
-    if not dojo.is_admin():
```
Comment (zardus) [inline dojo_plugin/pages/dojos.py:127]: Did this slip in by accident? Let's undo this, and then I think we're good to go!

**Review by zardus — CHANGES_REQUESTED** [review]
One last change :-)

**Comment by zardus** [comment]
Looks like a flaky failure... Retrying the tests...

**Comment by zardus** [comment]
Thank you!

**Comment by zardus** [comment]
So within 2 days of merging this, I almost deleted a dojo on the main instance by accident! Thank goodness for the confirm dialog.

Does your usecase _need_ this to be accessible via the web interface, or would it work for you if we move the trigger to the `dojo` utility (e.g., a `dojo delete $DOJO_ID` on the same level as `dojo db` and `dojo update`)?

**Comment by wumingzhilian** [comment]
> So within 2 days of merging this, I almost deleted a dojo on the main instance by accident! Thank goodness for the confirm dialog.
> 
> Does your usecase _need_ this to be accessible via the web interface, or would it work for you if we move the trigger to the `dojo` utility (e.g., a `dojo delete $DOJO_ID` on the same level as `dojo db` and `dojo update`)?

oh 
I previously saw SQL statements for deleting dojo on Discord. If placed in the dojo utility, it might not represent any change in managing the dojo compared to before. It would still require the execution of some commands, which is not very convenient. This delete function is something that admin need during the actual development and testing of the dojo, which is why I put it on the web interface.

However, if there is concern about data loss, it is actually possible to change the delete function to modify a 'delete' field in the database to 1 or 0. This might be a bit more troublesome.

**Pull request merged**